name: CI

on:
  push:
    branches: [master]
  pull_request: {}
  
permissions:
  contents: read

jobs:
  mypy:
    runs-on: ubuntu-20.04

    steps:
      - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
      - name: Set up Python
        uses: actions/setup-python@0a5c61591373683505ea898e09a3ea4f39ef2b9c # v5.0.0
      - name: Install dependencies
        run: pip install mypy
      - name: Run mypy
        run: mypy --strict certifi

  test:
    runs-on: ubuntu-20.04
    strategy:
      matrix:
        python-version:
        - "3.6"
        - "3.7"
        - "3.8"
        - "3.9"
        - "3.10"
        - "3.11"
        - "3.12-dev"
    steps:
    - uses: actions/checkout@9bb56186c3b09b4f86b1c65136769dd318469633 # v4.1.2
    - name: Set up Python ${{ matrix.python-version }}
      uses: actions/setup-python@0a5c61591373683505ea898e09a3ea4f39ef2b9c # v5.0.0
      with:
        python-version: ${{ matrix.python-version }}
    - name: Install test dependencies
      run: |
        python -m pip install --upgrade pip
        python -m pip install pytest
    - name: Test with pytest
      run: |
        python -W error -W 'ignore:Running attrs on Python 3.6' -m pytest
